// standardized-audio-context/build/es2018/factories/media-stream-audio-source-node-constructor.js
import { getNativeAudioContext } from '../helpers/get-native-audio-context';
const DEFAULT_OPTIONS = {
    channelCount: 2,
    channelCountMode: 'max',
    channelInterpretation: 'speakers'
};
export const createMediaStreamAudioSourceNodeConstructor = (createNativeMediaStreamAudioSourceNode, createNotSupportedError, isNativeOfflineAudioContext, noneAudioDestinationNodeConstructor) => {
    return class MediaStreamAudioSourceNode extends noneAudioDestinationNodeConstructor {
        constructor(context, options) {
            const nativeContext = getNativeAudioContext(context);
            if (isNativeOfflineAudioContext(nativeContext)) {
                throw createNotSupportedError();
            }
            const mergedOptions = Object.assign({}, DEFAULT_OPTIONS, options);
            const nativeMediaStreamAudioSourceNode = createNativeMediaStreamAudioSourceNode(nativeContext, mergedOptions);
            super(context, nativeMediaStreamAudioSourceNode, null);
            // Bug #63: Edge & Firefox do not expose the mediaStream yet.
            this._mediaStream = mergedOptions.mediaStream;
            this._nativeMediaStreamAudioSourceNode = nativeMediaStreamAudioSourceNode;
        }
        get mediaStream() {
            return (this._nativeMediaStreamAudioSourceNode.mediaStream === undefined) ?
                this._mediaStream :
                this._nativeMediaStreamAudioSourceNode.mediaStream;
        }
    };
};
